File: kickstart/2019ks_roundA_2.cpp
#include<bits/stdc++.h>
using namespace std;

const int maxn = 251;
const int des[4][2]={{0,1},{0,-1},{1,0},{-1,0}};

int n,m;
int vis[maxn][maxn];
char ch[maxn][maxn];

queue<int> qu;
bool che(int x){
    //cout<<"x = "<<x<<endl;
    for(int i=0;i<n;i++)for(int j=0;j<m;j++)vis[i][j]=-1;
    int cnt=0;
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(ch[i][j]=='1'){
                vis[i][j]=0;
                qu.push(i*m+j);
                cnt++;
            }
        }
    }
    while(qu.size()){
        int i = qu.front()/m,j = qu.front()%m;
        qu.pop();
        if(vis[i][j]==x)continue;
        for(int k=0;k<4;k++){
            int ni = i + des[k][0];
            int nj = j + des[k][1];
            if(ni>=0 && ni<n && nj>=0 && nj<m && vis[ni][nj]==-1){
                vis[ni][nj] = vis[i][j]+1;
                qu.push(ni*m+nj);
                cnt++;
            }
        }
    }
    if(cnt == n*m)return 1;
    int mx1=-10000000,mx2=-10000000;
    int mn1=10000000,mn2=10000000;
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++)if(vis[i][j]==-1){
            //cout<<"i,j = "<<i<<" "<<j<<endl;
            mx1 = max(mx1, i+j);
            mn1 = min(mn1, i+j);
            mx2 = max(mx2, i-j);
            mn2 = min(mn2, i-j);
        }
    }
    //cout<<mx1<<" "<<mn1<<endl;
    //cout<<mx2<<" "<<mn2<<endl;
    if(mx1-x<=mn1+x && mx2-x<=mn2+x){
        if(mn1+x>=mx2-x){
            if((mn1 + mx2)&1){
                if(mx1-x<mn1+x || mx2-x<mn2+x)return 1;
                else return 0;
            }
            else return 1;
        }
    }
    return 0;
}
int main()
{
    int t;scanf("%d",&t);
    for(int fir = 1; fir<=t; fir++){
        scanf("%d%d",&n,&m);
        for(int i=0;i<n;i++)scanf("%s",ch[i]);
        int l=0,r=m+n;
        while(l<r){
            int mid = l+r>>1;
            if(che(mid))r=mid;
            else l=mid+1;
        }
        printf("Case #%d: %d\n",fir,l);
    }
    return 0;
}

File: kickstart/2019ks_roundA_2.py
import queue

t = int(input())
fir = 0
while t>0:
    t -= 1
    fir += 1
    n, m = list(map(int, input().split()))
    ch = [0]*n
    for i in range(n):
        ch[i] = input()
    def che(x):
        global n
        global m
        global ch
        des = [[0,1],[0,-1],[1,0],[-1,0]]
        vis = [[0 for i in range(m)] for j in range(n)]
        qu = queue.Queue()
        ant = 0
        for i in range(n):
            for j in range(m):
                if ch[i][j] == '1':
                    qu.put([i,j,0])
                    vis[i][j] = 1
                    ant += 1
        while not qu.empty():
            i,j,now = qu.get()
            if now == x:
                continue
            for k in range(4):
                ni = i + des[k][0]
                nj = j + des[k][1]
                if ni>=0 and ni<n and nj>=0 and nj<m and vis[ni][nj] == 0:
                    vis[ni][nj] = 1
                    qu.put([ni,nj,now+1])
                    ant += 1
        if ant == n*m:
            return True
        mx1 = mx2 = -100000
        mn1 = mn2 = 100000
        
        for i in range(n):
            for j in range(m):
                if vis[i][j] == 0:
                    mx1 = max(mx1, i+j)
                    mn1 = min(mn1, i+j)
                    mx2 = max(mx2, i-j)
                    mn2 = min(mn2, i-j)
        if mx1-x<=mn1+x and mx2-x<=mn2+x:
            if mn1+x>=mx2-x:
                if (mn1+mx2)&1:
                    if mx1-x<mn1+x or mx2-x<mn2+x:
                        return True
                    else:
                        return False
                else:
                    return True
        return False
    

    che(1)
    l = 0
    r = m+n
    while l<r:
        mid = (l+r)>>1
        if che(mid):
            r = mid
        else:
            l = mid + 1
    print("Case #%d: %d" % (fir,l))
